lib/cfg-parser.c

```diff
@@ -186,68 +186,100 @@ CfgParser main_parser =
 
 CFG_PARSER_IMPLEMENT_LEXER_BINDING(main_, gpointer *)
 
+/* display CONTEXT lines before and after the offending line */
+#define CONTEXT 5
+
 static void
-_underline_source(YYLTYPE *yylloc, gchar *buf)
+_underline_source(YYLTYPE *yylloc, gchar **lines, gint error_index)
 {
-  gint i;
+  const gchar *buf;
+  gint line;
 
-  if (buf[0])
+  fprintf(stderr, "\n");
+  for (line = 0; lines[line] && line < error_index + CONTEXT + 1; line++)
     {
-      fprintf(stderr, "\n%s", buf);
+      buf = lines[line];
+
+      if (line == error_index)
+        fprintf(stderr, "%-8s", "-->");
+      else
+        fprintf(stderr, "%-8d", yylloc->first_line + line - error_index);
+
+      fprintf(stderr, "%s", buf);
       if (buf[strlen(buf) - 1] != '\n')
         fprintf(stderr, "\n");
-      for (i = 0; buf[i] && i < yylloc->first_column - 1; i++)
+
+      if (line == error_index)
         {
-          fprintf(stderr, "%c", buf[i] == '\t' ? '\t' : ' ');
+          /* do the underline */
+          fprintf(stderr, "%-8s", "");
+          for (gint i = 0; buf[i] && i < yylloc->first_column - 1; i++)
+            {
+              fprintf(stderr, "%c", buf[i] == '\t' ? '\t' : ' ');
+            }
+          for (gint i = yylloc->first_column; (i == yylloc->first_column) || (i < yylloc->last_column); i++)
+            fprintf(stderr, "^");
+          fprintf(stderr, "\n");
         }
-      for (i = yylloc->first_column; (i == yylloc->first_column) || (i < yylloc->last_column); i++)
-        fprintf(stderr, "^");
-      fprintf(stderr, "\n");
     }
 }
 
 static void
 _report_file_location(const gchar *filename, YYLTYPE *yylloc)
 {
   FILE *f;
-  gint lineno = 1;
+  gint lineno = 0;
   gchar buf[1024];
+  GPtrArray *context = g_ptr_array_new();
+  gint error_index = 0;
 
   f = fopen(filename, "r");
   if (f)
     {
-      while (fgets(buf, sizeof(buf), f) && lineno < yylloc->first_line)
-        lineno++;
-      if (lineno != yylloc->first_line)
-        buf[0] = 0;
+      while (fgets(buf, sizeof(buf), f) && lineno < (gint) yylloc->first_line + CONTEXT)
+        {
+          lineno++;
+          if (lineno > (gint) yylloc->first_line + CONTEXT)
```

Comment:
The while condition above ensures, `lineno < (gint) yylloc->first_line + CONTEXT`, together with lineno++ we can only have `lineno <= (gint) yylloc->first_line + CONTEXT`. I think this condition cannot be triggered, it can be safely deleted.